core/server/proxy/src/test/java/alluxio/proxy/s3/RateLimitInputStreamTest.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.proxy.s3;
+
+import static alluxio.Constants.KB;
+import static alluxio.Constants.MB;
+
+import com.google.common.util.concurrent.RateLimiter;
+import org.apache.commons.io.IOUtils;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.FutureTask;
+import java.util.stream.Collectors;
+
+public class RateLimitInputStreamTest {
+
+  private byte[] mData;
+
+  @Before
+  public void init() throws IOException {
+    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream(MB);
+    int count = 0;
+    while (count < MB) {
+      byte[] bytes = UUID.randomUUID().toString().getBytes();
+      byteArrayOutputStream.write(bytes);
+      count += bytes.length;
+    }
+    mData = Arrays.copyOf(byteArrayOutputStream.toByteArray(), MB);
+  }
+
+  @Test
+  public void testSingleThreadRead() throws IOException {
+    long rate = 100 * KB;
+    long globalRate = 200 * KB;
+    ByteArrayInputStream inputStream = new ByteArrayInputStream(mData);
+    RateLimitInputStream rateLimitInputStream = new RateLimitInputStream(inputStream,
+        RateLimiter.create(rate), RateLimiter.create(globalRate));
+    ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream(MB);
+    long start = System.currentTimeMillis();
+    IOUtils.copy(rateLimitInputStream, byteArrayOutputStream, KB);
+    long end = System.currentTimeMillis();
+    long duration = end - start;
+    long expectedDuration = MB / rate * 1000;
```

Comment:
can u also add an alternative case here where the first ratelimiter is higher then the second ratelimiter, and the expected rate should be limited by the smaller one?

Reply:
Yes, of course.